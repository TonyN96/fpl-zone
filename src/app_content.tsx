import React from "react";
import { useQuery } from "react-query";
import { BrowserRouter as Router,Route, Routes } from "react-router-dom";
import { useMediaQuery } from "@mui/material";
import { getAllFixtures, getGameData } from "api/fpl_api_provider";
import {
  AccountPage,
  AnalysisPage,
  FixturesAndResultsPage,
  GameweekLivePage,
  LoginPage,
  MyTeamPage,
  RegisterPage,
  ResetPage
} from "pages";
import PrivateRoute from "private_route";
import { AppData } from "types";

import { Logout } from "components/authentication";
import { Notifier, Startup } from "components/layout";

export const AppDataContext = React.createContext<AppData | null>(null);

export default function AppContent(): JSX.Element {
  // Fetching game data which will be made available throughout the app via context provider
  const {
    data: gameData,
    isError: gameDataIsError,
    error: gameDataError,
    isLoading: gameDataIsLoading
  } = useQuery("game-data", getGameData);

  // Fetching fixture data which will be made available throughout the app via context provider
  const {
    data: fixtureData,
    isError: fixtureDataIsError,
    error: fixtureDataError,
    isLoading: fixtureDataIsLoading
  } = useQuery("all-fixtures", getAllFixtures);

  const isCompact = useMediaQuery("(max-width:1500px)");

  const isLoading = gameDataIsLoading || fixtureDataIsLoading;
  const isError = gameDataIsError || fixtureDataIsError;
  const appData = gameData && fixtureData ? { ...gameData, fixtures: fixtureData, isCompact } : null;

  if (isLoading) {
    // Display loading message if data is still being fetched
    return (
      <Startup>
        <Notifier />
      </Startup>
    );
  } else if (isError) {
    // Display error message if data fetch failed
    const error = gameDataError || fixtureDataError;
    const errorMessage = error instanceof Error ? `: ${error.message}` : ".";

    return (
      <Startup>
        <Notifier type='error' message={`An error has occured: ${errorMessage}`} />
      </Startup>
    );
  }

  // Otherwise, render the app
  return (
    <AppDataContext.Provider value={appData}>
      <Router>
        <Routes>
          <Route path='*' element={<LoginPage />} />
          <Route path='/' element={<LoginPage />} />
          <Route path='/login' element={<LoginPage />} />
          <Route path='/register' element={<RegisterPage />} />
          <Route path='/reset' element={<ResetPage />} />
          <Route path='/logout' element={<Logout />} />
          <Route path='/account' element={<PrivateRoute component={<AccountPage />} />} />
          <Route
            path='/gameweek-live'
            element={<PrivateRoute component={<GameweekLivePage />} />}
          />
          <Route path='/my-team' element={<PrivateRoute component={<MyTeamPage />} />} />
          <Route
            path='/fixtures-and-results'
            element={<PrivateRoute component={<FixturesAndResultsPage />} />}
          />
          <Route path='/analysis' element={<PrivateRoute component={<AnalysisPage />} />} />
        </Routes>
      </Router>
    </AppDataContext.Provider>
  );

}
