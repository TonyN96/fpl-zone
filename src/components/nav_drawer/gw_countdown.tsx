import React from "react";
import { getGameData } from "api/fpl_api_provider";
import { useQuery } from "react-query";
import { Gameweek } from "types/gameweek";
import Countdown from "react-countdown";
import { Box, Typography } from "@mui/material";

export default function GameweekCountdown(): JSX.Element {
  const { data, isError } = useQuery("game-data", getGameData);

  if (isError || !data) return <></>;

  const gameweeks: Gameweek[] = data.events;
  const nextGameweek: Gameweek | undefined = gameweeks.find((gw) => {
    return gw.is_next;
  });

  if (!nextGameweek) return <></>;

  const deadline = nextGameweek.deadline_time;

  interface CountdownRenderProps {
    days: number;
    hours: number;
    minutes: number;
    completed: boolean;
  }

  const renderer = ({ days, hours, minutes, completed }: CountdownRenderProps): JSX.Element => {
    const gameweekName = nextGameweek.name;
    const daysLabel = days ? days + " day" + (days > 1 ? "s" : "") : "";
    const hoursLabel = hours ? hours + " hr" + (hours > 1 ? "s" : "") : "";
    const minsLabel = minutes ? minutes + " min" + (minutes > 1 ? "s" : "") : "";
    const countdown = [daysLabel, hoursLabel, minsLabel].join(" ");
    if (completed) {
      return (
        <Typography>
          {gameweekName}
          <br />
          in progress
        </Typography>
      );
    } else {
      return (
        <Typography>
          {gameweekName} deadline:
          <br />
          {countdown}
        </Typography>
      );
    }
  };

  return (
    <Box textAlign="center" sx={{ pb: 2 }}>
      <Countdown date={deadline} renderer={renderer} />
    </Box>
  );
}
