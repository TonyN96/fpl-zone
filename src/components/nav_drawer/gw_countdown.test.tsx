import React from "react";
import { render, screen } from "@testing-library/react";
import { AppDataContext } from "app_content";
import moment from "moment";
import { mockAppData } from "test";

import "@testing-library/jest-dom/extend-expect";

import { GameweekCountdown } from "./gw_countdown";

describe("Gameweek countdown tests", () => {
  const createComponent = (): JSX.Element => {
    return (
      <AppDataContext.Provider value={mockAppData}>
        <GameweekCountdown />
      </AppDataContext.Provider>
    );
  };

  it("Snapshot test", () => {
    const { asFragment } = render(createComponent());
    expect(asFragment()).toMatchSnapshot();
  });

  const setGameweekDeadline = (days = 0, hours = 0, minutes = 0): void => {
    const currentDate = moment();
    const mockDeadlineDate = currentDate.add({ days, hours, minutes }).toISOString();
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const nextGameweek = mockAppData.gameweeks.find(gw => gw.is_next)!;
    nextGameweek.deadline_time = mockDeadlineDate;
  };

  const nextGameweekName = "GAMEWEEK 34 DEADLINE";

  it("Days, hours and minutes", () => {
    setGameweekDeadline(2, 16, 27);
    render(createComponent());

    const expectedCountdownText = "2 DAYS 16 HRS 27 MINS";
    const gameweekDeadlineTime = screen.getByTestId("gameweek-deadline-container");

    expect(gameweekDeadlineTime).toHaveTextContent(nextGameweekName);
    expect(gameweekDeadlineTime).toHaveTextContent(expectedCountdownText);
  });

  it("Hours and minutes only", () => {
    setGameweekDeadline(0, 3, 52);
    render(createComponent());

    const expectedCountdownText = "3 HRS 52 MINS";
    const gameweekDeadlineTime = screen.getByTestId("gameweek-deadline-container");

    expect(gameweekDeadlineTime).toHaveTextContent(nextGameweekName);
    expect(gameweekDeadlineTime).toHaveTextContent(expectedCountdownText);
  });

  it("Minutes only", () => {
    setGameweekDeadline(0, 0, 12);
    render(createComponent());

    const expectedCountdownText = "12 MINS";
    const gameweekDeadlineTime = screen.getByTestId("gameweek-deadline-container");

    expect(gameweekDeadlineTime).toHaveTextContent(nextGameweekName);
    expect(gameweekDeadlineTime).toHaveTextContent(expectedCountdownText);
  });

  it("Singular times", () => {
    setGameweekDeadline(1, 1, 1);
    render(createComponent());

    const expectedCountdownText = "1 DAY 1 HR 1 MIN";
    const gameweekDeadlineTime = screen.getByTestId("gameweek-deadline-container");

    expect(gameweekDeadlineTime).toHaveTextContent(nextGameweekName);
    expect(gameweekDeadlineTime).toHaveTextContent(expectedCountdownText);
  });

  it("Displays \"Gameweek in progress\" if duration has elapsed", () => {
    setGameweekDeadline();
    render(createComponent());

    const expectedCountdownText = "IN PROGRESS";
    const gameweekDeadlineTime = screen.getByTestId("gameweek-deadline-container");

    expect(gameweekDeadlineTime).toHaveTextContent(expectedCountdownText);
  });
});
