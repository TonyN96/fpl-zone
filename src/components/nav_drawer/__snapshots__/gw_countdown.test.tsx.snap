// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gameweek countdown tests Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-oevum5"
    data-testid="gameweek-deadline-container"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
    >
      GAMEWEEK 34
      <br />
      IN PROGRESS
    </h5>
  </div>
</DocumentFragment>
`;
